multipass-addressing/src/main/scala/com/precisely/bigdata/sample/spark/MultipassAddressing.scala

```diff
@@ -0,0 +1,159 @@
+    val dataLocation = args(2)
+    val downloadLocation = args(3)
+    val outputDirectory = args(4)
+
+    val sparkConf = new SparkConf()
+    sparkConf.setIfMissing("spark.master", "local[*]")
+    val session = SparkSession.builder()
+      .config(sparkConf)
+      .getOrCreate()
+
+    // Load the addresses from the csv
+    val addressInputDF = AddressInput.open(session, inputAddressPath)
+
+    // set preference to return all fields
+    val prefBuilder = new PreferencesBuilder().withReturnAllInfo(true).build()
+
+    val downloadManager = new DownloadManagerBuilder(downloadLocation)
+      .addDownloader(new S3Downloader(session.sparkContext.hadoopConfiguration))
+      .addDownloader(new HDFSDownloader(session.sparkContext.hadoopConfiguration))
+      .addDownloader(new LocalFilePassthroughDownloader())
+      .build()
+
+    // build a AddressingUDF, with custom executor
+    val addressingUdf: UserDefinedFunction = new AddressingBuilder()
+      .withResourcesLocation(resourcesLocation)
+      .withDataLocations(dataLocation)
+      .withDownloadManager(downloadManager)
+      .udfBuilder()
+      .withPreferences(prefBuilder)
+      .withOutputFields(
+        "location.feature.geometry.coordinates.x as X",
+        "location.feature.geometry.coordinates.y as Y",
+        "address.formattedStreetAddress as formattedStreetAddress",
+        "address.formattedLocationAddress as formattedLocationAddress",
+        "customFields['PRECISION_CODE'] as PRECISIONCODE")
+      .withErrorField("error")
+      .forCustomExecutor(new CustomExecutor())
+
+    // call UDF for each row in the address DataFrame, this will result in the dataframe containing a new column for each
+    // of the specified output fields as well as an error column
+    addressInputDF
+      // Adds a new column, represented as a collection comprised of the outputFields and the error field
+      .withColumn("addressing_geocode_result", addressingUdf(map(
+        lit("street"), col("streetName"),
+        lit("city"), col("city"),
+        lit("admin1"), col("state"),
+        lit("postalCode"), col("zip")))
+      )
+      // Persist the addressing result to avoid recalculation when we expand the result
+      .persist()
+      // Expand the result collection such that each output field is a separate column, including the error field.
+      .select("*", "addressing_geocode_result.*").drop("addressing_geocode_result")
+      // Write the dataframe to the specified output folder as parquet file
+      .write.mode(SaveMode.Overwrite).option("header","true").csv(outputDirectory)
+  }
+}
+
+/*
+ * A custom Executor that contains the logic to perform verify and then multipass geocode
+ */
+class CustomExecutor extends com.precisely.bigdata.addressing.spark.api.AddressingExecutor {
+
+  import com.precisely.addressing.v1.model.Response
+  import com.precisely.addressing.v1.{Addressing, Preferences}
+
+  var relaxedGeocoderPreferences: Preferences = _
```

Comment:
change this from a 'var' to a 'private val', moving the preference builder calls here.

Reply:
Ok